File: php/src/Common/Infrastructure/Stream/Producer/SerializableEvent.php
<?php
declare(strict_types=1);

namespace Demo\App\Common\Infrastructure\Stream\Producer;

readonly abstract class SerializableEvent
{
    public function __construct(
        public string $id,
        public string $schema,
        public string $eventType,
        public string $version,
        public \DateTimeImmutable $occurredOn,
        public array $payload,
    ) {}

    public function toJson(): string
    {
        return json_encode([
            'id' => $this->id,
            'schema' => $this->schema,
            'eventType' => $this->eventType,
            'version' => $this->version,
            'occurredOn' => $this->occurredOn->format(\DateTime::ATOM),
            'payload' => json_encode($this->payload),
        ]);
    }
}

File: kotlin/src/main/kotlin/common/infrastructure/stream/producer/SerializableEvent.kt
package common.infrastructure.stream.producer

import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

abstract class SerializableEvent(
    open val id: String,
    open val schema: String,
    open val eventType: String,
    open val version: String,
    open val occurredOn: LocalDateTime,
    open val payload: Map<String, Any>
) {
    fun toJson(): String {
        return """
            {
                "id": "$id",
                "schema": "$schema",
                "eventType": "$eventType",
                "version": "$version",
                "occurredOn": "${occurredOn.format(DateTimeFormatter.ISO_DATE_TIME)}",
                "payload": ${payloadToJson()}
            }
        """.trimIndent()
    }

    private fun payloadToJson(): String {
        val payloadJson = payload.map { (key, value) ->
            "\"$key\": \"$value\""
        }.joinToString(", ")

        return "{ $payloadJson }"
    }
}
